tests/cpp_new/analysis/alias_analysis_test.cpp

```diff
@@ -0,0 +1,119 @@
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "taichi/ir/analysis.h"
+#include "taichi/ir/ir_builder.h"
+#include "taichi/ir/snode.h"
+#include "taichi/ir/statements.h"
+
+namespace taichi {
+namespace lang {
+namespace irpass {
+namespace analysis {
+namespace {
+
+std::vector<Stmt *> make_const_indices(const std::vector<int> indices,
+                                       IRBuilder *builder) {
+  std::vector<Stmt *> result;
+  result.reserve(indices.size());
+  std::for_each(indices.begin(), indices.end(), [builder, &result](int i) {
+    result.push_back(builder->get_int32(i));
+  });
+  return result;
+}
+
+}  // namespace
+
+TEST(AliasAnalysis, Alloca) {
+  IRBuilder builder;
+
+  auto *alloca1 = builder.create_local_var(PrimitiveType::f32);
+  auto *alloca2 = builder.create_local_var(PrimitiveType::f32);
+
+  auto aa = alias_analysis(alloca1, alloca1);
+  EXPECT_EQ(aa, AliasResult::same);
+  aa = alias_analysis(alloca1, alloca2);
+  EXPECT_EQ(aa, AliasResult::different);
+}
+
+TEST(AliasAnalysis, GlobalPtr_Same) {
+  auto snode = std::make_unique<SNode>(/*depth=*/1, /*t=*/SNodeType::place);
+
+  IRBuilder builder;
+  const auto indices = make_const_indices({1, 2, 3}, &builder);
+  auto *gptr1 = builder.create_global_ptr(snode.get(), indices);
+  auto *gptr2 = builder.create_global_ptr(snode.get(), indices);
+
+  const auto aa = alias_analysis(gptr1, gptr2);
+  EXPECT_EQ(aa, AliasResult::same);
+}
+
+TEST(AliasAnalysis, GlobalPtr_DiffIndices) {
+  auto snode = std::make_unique<SNode>(/*depth=*/1, /*t=*/SNodeType::place);
+
+  IRBuilder builder;
+  const auto indices1 = make_const_indices({1, 2, 3}, &builder);
+  const auto indices2 = make_const_indices({1, 3, 2}, &builder);
+  auto *gptr1 = builder.create_global_ptr(snode.get(), indices1);
+  auto *gptr2 = builder.create_global_ptr(snode.get(), indices2);
+
+  const auto aa = alias_analysis(gptr1, gptr2);
+  EXPECT_EQ(aa, AliasResult::different);
+}
+
+TEST(AliasAnalysis, GlobalPtr_DiffIndicesWithinTrailingBits) {
+  auto snode = std::make_unique<SNode>(/*depth=*/1, /*t=*/SNodeType::place);
+  for (int i = 0; i < 2; ++i) {
+    snode->physical_index_position[i] = i;
+  }
+  snode->extractors[snode->physical_index_position[2]].trailing_bits = 2;
+
+  IRBuilder builder;
+  // The difference along the 2-dimension is (6 -3) = 3, which is smaller than
+  // (1 << trailing_bits) = 4.
+  const auto indices1 = make_const_indices({1, 2, 3}, &builder);
+  const auto indices2 = make_const_indices({1, 2, 6}, &builder);
```

Comment:
Well... IIUC `a[1, 2, 3]` and `a[1, 2, 6]` are **different** elements when `trailing_bits == 2`.

Maybe we should change
```C++
      if (!diff.first) {
        uncertain = true;
      } else if (std::abs(diff.second) >=
                 (1 << snode->extractors[snode->physical_index_position[i]]
                           .trailing_bits)) {
        return AliasResult::different;
      }
```
to
```C++
      if (!diff.first || diff.second != 0) {
        uncertain = true;
      }
      if (std::abs(diff.second) >=
          (1 << snode->extractors[snode->physical_index_position[i]]
                           .trailing_bits)) {
        return AliasResult::different;
      }
```

Reply:
Hmm, could you explain the original intention when writing this in https://github.com/taichi-dev/taichi/pull/1657?

Reply:
Thanks, fixed